Miner backport

kyuupichan opened the pull request:
Ultimate goal here is a backport of CPFP. That depends on various prior changes to miner.cpp/h that we don't have backported to BU yet. This series of commits are backports of 2 Core PRs that take us right up to the point where CPFP related PRs start to go in.

PR bitcoin#7539
PR bitcoin#7598

This is obviously very sensitive code given the recent big block that was generated. I've done my best to preserve existing semantics except for one case. The old code had a 16 byte reserve for extra varints in the coinbase.  I've replaced it; there were 3 incorrect assumptions:

1) The height in the coinbase isn't a varint, it's a script serialized integer, which is a minimum of 2 bytes and currently 4 bytes (for heights of 32768 and higher) and will be 4 bytes for another 150 years.  See BIP34
2) The coinbase output amount isn't a varint, it's always 8 bytes like all txout values.
3) There was no accounting for the txCount varint in the serialized block.  I reserve 5 bytes for 4 billion txs, which should be enough for a while :)  Current full blocks will only be using 3 bytes, but that will bump to 5 once blocks can hold 65536 txs.

The test loops are more fine-grained, and assert that blocks up to and including a tighter limit (because of the reasons above, such as a 2-byte height of 110) are created, but no greater.  Comments are included that explain the limits.  This gives me a high degree of confidence my understanding and implementation is correct.

Nevertheless careful review appreciated @gandrewstone and @ptschip

morcos committed:
Refactor CreateNewBlock to be a method of the BlockAssembler class

sdaftuar committed:
Add tags to mempool's mapTx indices

morcos committed:
FIX: Account for txs already added to block in addPriorityTxs

morcos committed:
FIX: correctly measure size of priority block

Commit:
Tweak tests

Commit:
Apply clang-format changes to modified code only

kyuupichan commented:
Rebased, and clang-format applied to the modified source (only)

gandrewstone commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
+    // This serializes with output value, a fixed-length 8 byte field, of zero and height, a serialized CScript
+    // signed integer taking up 4 bytes for heights 32768-8388607 (around the year 2167) after which it will use 5.
+    nCoinbaseSize = coinbaseTx(scriptPubKeyIn, 400000, 0).GetSerializeSize(SER_NETWORK, PROTOCOL_VERSION);
+    // BU005 END
```
is this comment irrelevant now?

kyuupichan replied:
You mean the BU005?  I left it in; I had no idea what it was referring to.

gandrewstone replied:
yes its a merge note... so can be removed

gandrewstone commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
+        // If tx already in block, skip
+        if (inBlock.count(iter))
+        {
+            assert(false); // shouldn't happen for priority txs
```
If somehow the same transaction is in the priority heap twice, we don't want to abort bitcoind, so let's change this assert(false) to DbgAssert(false, );
(don't put the continue in the DbgAssert, loop control does not work inside it).

gandrewstone commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
-                iter = clearedTxs.top();
-                clearedTxs.pop();
-            }
+            iter = mempool.mapTx.project<0>(mi);
```
can iter ever equal end() here?

kyuupichan replied:
The loop logic is hard to read, for sure.  iter can only be end() if mi is also end() as one is a projection of the other on the same underlying mempool tx.  The loop condition implies that to enter the loop, if mi is at the end, then clearedTxs must be non-empty.  Since the line in question is for when clearedTxs is empty, the answer is no.

gandrewstone commented:
ok I'm ready to merge, except for those three small issues.

ptschip commented:
@gandrewstone   looks good to me

zander commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
-    if(!pblocktemplate.get())
-        return NULL;
-    CBlock *pblock = &pblocktemplate->block; // pointer for convenience
+    CBlockTemplate *tmpl = NULL;
```
not NULL,  use nullptr.

kyuupichan replied:
Fixed

zander commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
-        return NULL;
-    CBlock *pblock = &pblocktemplate->block; // pointer for convenience
+    CBlockTemplate *tmpl = NULL;
+    if (maxGeneratedBlock > BLOCKSTREAM_CORE_MAX_BLOCK_SIZE)
```
Why are you using global maxGeneratedBlock instead of the class member nBlockMaxSize ?

kyuupichan replied:
Prior code did.  I agree nBlockMaxSize is better; I've updated it.

zander commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
-    txNew.vout.resize(1);
-    txNew.vout[0].scriptPubKey = scriptPubKeyIn;
+    // If the block is too small we need to drop back to the 1MB ruleset
+    if ((!tmpl) || (tmpl->block.nBlockSize <= BLOCKSTREAM_CORE_MAX_BLOCK_SIZE))
```
This comment makes no sense, the size of the block doesn't determine the ruleset, the fact that a > 1MB block is on the block chain is the determining factor.

Re-doing your entire block is also a bad idea, as the *only* change is the checking of the block. Why don't you check it instead. And redo it only if if fails the conservative check?

zander commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
+    nLastBlockTx = nBlockTx;
+    nLastBlockSize = nBlockSize;
+    LogPrintf(
+        "CreateNewBlock(): total size %u txs: %u fees: %ld sigops %d\n", nBlockSize, nBlockTx, nFees, nBlockSigOps);
```
nBlockSigOps is unsigned, you pass it as %d. Thats incorrect.

Use just 'printf' instead of 'LogPrintf' and compile to get all warnings from gcc, then when you fixed the warnings replace the LogPrintf call.

kyuupichan replied:
Fixed.

zander commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
-    CAmount nFees = 0;
 
+    uint64_t nTxSigOps = iter->GetSigOpCount();
+    if (nBlockSize + nTxSize <= BLOCKSTREAM_CORE_MAX_BLOCK_SIZE) // Enforce the "old" sigops for <= 1MB blocks
```
The actual block size is not an indication of the rules, so this is incorrect, just like above.

The fact if there has been a >1MB block on the chain before is the only indication.

kyuupichan replied:
I will defer this one too

zander commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
 }
 
-CBlockTemplate* CreateNewBlock(const CChainParams& chainparams, const CScript& scriptPubKeyIn,bool blockstreamCoreCompatible);
+BlockAssembler::BlockAssembler(const CChainParams &_chainparams) : chainparams(_chainparams)
```
Please initialize all your members in the constructor.

pblock, nBlockSize, nBlockTx, nBlockSigOps, nFees, nHeight, nLockTimeCutoff, lasFewTxs, blockFinished were all forgotten here.

I didn't check very carefully, but are you sure all of those are actually useful at class scope? I have the strong impression half of them can be limited to only exist in the method they were used.

Please also remove the CChainParams instance, just use the global static method Params() when you need it.

kyuupichan replied:
Hey Tom, thanks for the careful look-over.  As this PR implies, I'm not attempting to improve the design here; I'm backporting a code redesign (which I agree is quite poor) that core did that led to CPFP with a view to getting CPFP in BU.  I'm happy to come back and clean up the design if that's what @gandrewstone wants, but I think trying to improve the poor state of affairs at the same time as backporting and merging in existing BU changes is too much churn in one go.

zander replied:
Importing code that doesn't initialize its members is not "cleanup" in my opinion, its very much a bug being imported.

zander commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
+    std::string cbmsg = FormatCoinbaseMessage(BUComments, minerComment);
+    const char *cbcstr = cbmsg.c_str();
+    vector<unsigned char> vec(cbcstr, cbcstr + cbmsg.size());
+    COINBASE_FLAGS = CScript() << vec;
```
Why is COINBASE_FLAGS all capital and why is it a global? I think it would be great to have as a member of the BlockAssembler class.

It should definitely not be all capital as that is reserved for global static const variables, and you are modifying it here, so its obviously not const.

kyuupichan replied:
I'm preserving the prior code.  I agree it could be improved, I think that should be done separately (either after this or after CPFP to reduce conflicts).

zander commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
+        uint64_t blockMbSize = 1 + ((nBlockSize + nTxSize - 1) / 1000000);
+        if (nBlockSigOps + nTxSigOps > blockMiningSigopsPerMb.value * blockMbSize)
+        {
+            if (nBlockSigOps > blockMiningSigopsPerMb.value * blockMbSize - 2)
```
This if() makes little sense, if you are exactly 1 sigop from the limit, you try again with a different transaction hoping to get better luck? That hard coded limit is too small to be relevant.  This is essentially a no-op.

kyuupichan replied:
Sorry don't understand your point here

gandrewstone replied:
this is preserving behavior in the original code.  You can see the same thing on line 286, and you'll probably find it in classic if you look.  And yes it does try to find a 1 sig-op tx.  Why not?  worst case you just go through the whole mempool

zander commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
+
+bool BlockAssembler::isStillDependent(CTxMemPool::txiter iter)
+{
+    BOOST_FOREACH (CTxMemPool::txiter parent, mempool.GetMemPoolParents(iter))
```
Notice that GetMemPoolParents asserts if iter is invalid, I've seen very little reason to think that code calling this avoids iter being invalid.

kyuupichan replied:
I think the iter passed from around line 367 is valid for the reasons I explained to @gandrewstone in response to his question.  From around line 445 vecPriority is just a heap of the mempool txs, and the loop condition guarantees it's not empty.

zander commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
+    resetBlock(scriptPubKeyIn);
+
+    pblocktemplate.reset(new CBlockTemplate());
+    if (!pblocktemplate.get())
```
This check is a no-op. Your new would have thrown an exception if it failed, and as such your unique pointer is never null here.

kyuupichan replied:
Indeed, removed.

zander commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
+                strprintf("%s: TestBlockValidity failed: %s", __func__, FormatStateMessage(state)));
+        }
+    }
+    if (pblock->fExcessive)
```
Why is fExcessive possible to be set without the TestBlockValidity returning false?  If that was fixed, you would not need this extra check.

kyuupichan replied:
I'll leave this one for @gandrewstone too

gandrewstone replied:
because an excessive block is valid in other contexts, such as when blocks are received.  Its invalid during mining because we don't want to produce a block that our own node will orphan.

zander commented on src/miner.cpp:
```diff
@@ -74,321 +74,406 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
+{
+    resetBlock(scriptPubKeyIn);
+
+    pblocktemplate.reset(new CBlockTemplate());
```
You have a unique pointer here with pblocktemplate. One which you fill in the start of this method and release at the end.
The problem is that you throw exceptions which avoids releasing this variable.

Why is this not a variable local to this method instead of a class member?

kyuupichan replied:
Your third sentence I think means "if an exception is thrown the block template isn't released"?  Agreed.

However I can't easily make it local as it's used elsewhere (AddToBlock).  I'll try and come up with an improvement here.

zander reviewed (commented):
I found quite a number of problems in this code that should be addressed, many of them actual bugs.

Commit:
Improvements based on PR feedback

- Initialize all members, remove member block
- make the member reference first so the init list is naturally well-ordered
- Remove old comment
- Use nullptr.  Don't do redundant null-point check
- make pBlockTemplate a local unique_ptr.  Pass a pointer to the
  template where necessary
- fix LogPrintf format specifiers
